### PR TITLE
oslayer/xkeyboardcontrol: add Unicode support

### PR DESCRIPTION
**Pull request by benoit-pierre**
This is done by dynamically changing the keymap when a character we want
to send is not already mapped. A rolling list of unused keycodes is used
to ensure we don't reuse a keycode before the target application has
time to process each sequence of keymap change + key press/release
events. The keymap entries used by Plover are marked with a special
keysym, so if Plover is killed/restarted, those entries are still
considered unused, and consequently reusable.

### CODE REVIEW

**Comment by morinted** [comment]
What flavors of Linux have you tried this on?

**Comment by benoit-pierre** [comment]
Arch Linux.

**Comment by morinted** [comment]
We should try on Ubuntu (don't care about the version) as I imagine that's what most people are using, just to see if there are any gotchyas.

**Comment by benoit-pierre** [comment]
Only standard X11 calls, so I don't expect any issue on another distribution. I did however think of some use case that could be problematic. Will try to test it and amend the code accordingly.

**Comment by benoit-pierre** [comment]
OK, so I've completely rewritten the code, and tested with a Ubuntu VM too.

File: plover/oslayer/xkeyboardcontrol.py [hunk 1]
```diff
@@ -246,18 +252,905 @@ def ignore_key_events(self, key_events):
...
+
+    def _update_keymap(self):
+        '''Analyse keymap, build a mapping of keysym to (keycode + modifiers),
+        and find unused keycodes that can be used for unmapped keysyms.
+        '''
+        self.keymap = {}
+        self.custom_mappings_queue = []
+        # Analyse X11 keymap.
+        keycode = self.display.display.info.min_keycode
+        nb_keycodes = self.display.display.info.max_keycode - keycode + 1
```
Comment (jeremy-w) [inline plover/oslayer/xkeyboardcontrol.py]: It took me a while to work out that `nb` was "number of". Maybe `keycode_count` would avoid the need to abbreviate?

Comment (benoit-pierre) [inline plover/oslayer/xkeyboardcontrol.py]: That's something I use a lot, `nb_something`, but I now realize that might be a mistake... How about: `num_keycodes`?

Comment (jeremy-w) [inline plover/oslayer/xkeyboardcontrol.py]: That would normally be clear, but this case is special, because we're dealing with keycodes. What's confusing with any identifier name using just "num" here is that it could be abbreviating "numeric". Presumably there are keycodes that correspond to the number keys, and those would be either "number keycodes" or "numeric keycodes", and then I'm confused again. The code clears it up, but it's a confusion we can avoid, either by spelling out `number_of_keycodes` or flipping it around to `keycode_count`.

Comment (benoit-pierre) [inline plover/oslayer/xkeyboardcontrol.py]: OK, will amend.

**Comment by benoit-pierre** [comment]
I'll rewrite the history once we're ready to merge.

**Comment by morinted** [comment]
Sounds good. I'm feeling as if you should merge when you think it's ready, as no one else can really test it.

If I were to test this, I suppose that I'd try:

- A normal sentence.
- A long word (supercalifragilisticexpialidocious).
- A command sequence (control c, control v, Alt tab).
- Arrow keys through a document.

Or, more simply:

```json
{
"TEFT": "The quick brown fox jumps over the lazy dog. Supercalifragilisticexpialidocious. {#Shift_L(Left Left Left) Control_L(c v v v v v). You should see 5 's. 's. All is well? All is well. 🐱 😆 💩 👪 ⁉." 
}
```

**Comment by benoit-pierre** [comment]
You're missing a `}`. It works fine. It's also unlikely to catch some of the corner cases. My test dictionaries is:

```
{
"-F":"♜♞♝♛♚♝♞♜",
"*F":"ǐ",
"-R":"♖♘♗♕♔♗♘♖",
"*R":"ħĦ7ðÐ8*þÞ×÷äÄåÅãÃøØđĐłŁúÚüÜöÖ«‹»",
"-P": "По оживлённым берегам",
"*P": "ПО ОЖИВЛЁННЫМ БЕРЕГАМ",
"-B": "Ἰοὺ ἰού· τὰ πάντʼ ἂν ἐξήκοι σαφῆ",
"*B": "ἸΟῪ ἸΟΎ· ΤᾺ ΠΆΝΤʼ ἊΝ ἘΞΉΚΟΙ ΣΑΦῆ",
"-L": "पशुपतिरपि तान्यहानि कृच्छ्राद् ",
"-G": "子曰：「學而時習之，不亦說乎？有朋自遠方來，不亦樂乎",
"-T": "ஸ்றீனிவாஸ ராமானுஜன் ஐயங்கார்",
"-S": "إِيَّاكَ نَعْبُدُ وَإِيَّاكَ نَسْتَعِينُ",
"-D": "♈♉♊♋♌♍♎♏♐♑♒♓",
"-Z": "àáâãäāăąȧǎȁȃèéêẽëēĕęėǎěȅȇìíîĩïīĭįiìǐȉȋòóôõöōŏǫȯőǒȍȏùúûũüūŭųǎěűǔȕȗỳýŷỹÿȳǎěűẏǎěűẏǎěűẏ",
"*Z": "ÀÁÂÃÄĀĂĄȦǍȀȂÈÉÊẼËĒĔĘĖǍĚȄȆÌÍÎĨÏĪĬĮIÌǏȈȊÒÓÔÕÖŌŎǪȮŐǑȌȎÙÚÛŨÜŪŬŲǍĚŰǓȔȖỲÝŶỸŸȲǍĚŰẎǍĚŰẎǍĚŰẎ"
}
```

But the problem I had thought about was with this:

```python
#!/usr/bin/env python2
# -*- coding: utf-8 -*-

import time

from plover.oslayer.xkeyboardcontrol import KeyboardEmulation
from plover import log


log.set_level(log.DEBUG)
kbd = KeyboardEmulation()

time; time.sleep(3)

kbd.send_string(u'ⅰⅱⅲⅳⅴⅵⅶⅷⅰ♜')
```

With only 8 free mappings, with the previous version of the code, the 9th character "ⅰ" would be wrong because for sending "♜" we reuse the same mapping.

The thing is, if you don't have enough free mappings and/or the application you send output is too slow, you can still run into issues because you'll reuse a mapping too soon... Like with only 4 free mappings and GMail in Firefox. (But for reference, I get 28 unused keycodes, so 56 free mappings with the standard colemak layout)

**Comment by morinted** [comment]
Well, some support is better than no support. I'm in favor of merging this now, same as the Windows rewrite, then at least we'll get early feedback when we release the first pre-release.

**Comment by benoit-pierre** [comment]
Agreed.
